Added module to find Launch Configurations

fiunchinho opened the pull request:
Added module to find Auto Scaling Group Launch Configurations in AWS. It let the user search by name, using regular expressions.

robynbergeron commented:
Hi @fiunchinho -- Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

gregdek commented:
Thanks @fiunchinho for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

fiunchinho commented:
I'll try to fix it, but faster feedback would be awesome. I opened this PR 7 months ago :cry:

gregdek commented:
@fiunchinho A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

fiunchinho commented:
Fixed problems and rebased history. Feedback will be appreciated

gregdek commented:
@fiunchinho Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

fiunchinho commented:
ready_for_review

gregdek commented:
Thanks @fiunchinho for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @fiunchinho for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@fiunchinho A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

fiunchinho commented:
ready_for_review

gregdek commented:
Thanks @fiunchinho for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

fiunchinho commented:
You can use this module like

```
---

- name: "Find old Launch Configs"
  ec2_lc_find:
    profile: "{{ boto_profile }}"
    region: "{{ aws_region }}"
    name_regex: "*nameToFind*"
    sort: true
    sort_end: -10
  register: old_launch_config

- name: "Remove old Launch Configs"
  ec2_lc:
    profile: "{{ boto_profile }}"
    region: "{{ aws_region }}"
    name: "{{ item.name }}"
    state: absent
  with_items: old_launch_config.results
  ignore_errors: yes
```

rgomezcasas commented:
works_for_me

ismferd commented:
works_for_me

fiunchinho commented:
shipit

tchia04 commented:
How do you suppress the big blob returned when the ec2_lc_find module is ran? 

How do you display just the launch configuration name using debug or other methods?

fiunchinho commented:
Can you showed me the code that you are trying to execute?

tchia04 commented:
Never mind. I was using an old version of ec2_lc_find that doesn't match the content of this pull request even though it had the same name.

gregdek commented:
OK, after review in the new modules meeting of August 10th, we've decided that some revisions are in order. Will leave the appropriate comments in a subsequent comment.

Thanks @fiunchinho for your patience. This module is in the "final review" process, so we'll look at it again every week.

nitzmahone commented:
Sorry @fiunchinho for the slow feedback- if you can work on/discuss the following, then CC me on further updates, I'll make sure we get it looked at ASAP and merged once we're both happy with the result.
- Doesn't look like you support paging, so anything with more than 100 launch configs will get lost. You should either error at 100 results (if you don't feel like using boto's paging) so the problem isn't silent, or add paging support to fetch all relevant results before returning.
- The various sort parameters are a little obtuse, and make for a somewhat unfriendly UI. I suggest the following: remove the "sort" arg altogether, and always default to an ascending sort (I can't think of any reason for unsorted values). Leave "sort_order" alone to allow for descending sort to be specified. Finally the sort_start/sort_end are rather strange- I can't see how those would be useful in a playbook when you don't know the exact series of launch configurations. I'd recommend removing, or coming up with some other scheme to filter them- the server-side regex seems good enough for that, though. 

Thanks!

fiunchinho commented:
Thank you for the feedback!

I'll drop the sort parameter, since you are completely right.
About `sort_start` and `sort_end`, they are useful when you want to fetch only the first/last N results. For example, if you want to [keep only the most 10 recent launch configurations and deleting the rest](https://github.com/ansible/ansible-modules-extras/pull/1023?utm_content=buffer9b70e&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer#issuecomment-234788087), you can use `sort_end` to fetch all the launch configurations but the last 10.

You could get similar results using `sort_start`. Maybe only one of them should be kept (renamed to `limit`?), but I wouldn't remove both.

What do you think @nitzmahone ?

nitzmahone commented:
Is that maybe the reason for the unsorted, then? I'm just having a hard time seeing why you'd want to return/drop launch configs positionally rather than by name/function...

fiunchinho commented:
Isn't wanting to keep only the latest N launch configuration enough?

nitzmahone commented:
If that were the case, wouldn't you need explicit "date_ascending/date_descending" sort modes too (boto doesn't define the returned order)? Again, I'm just struggling to see viable use cases for sort_start/end args as it's built right now... The "expensive" part is fetching all the launch descriptions from boto to sort/filter (since boto doesn't do the sorting/filtering itself, so you have to pull them all to the client), not shipping them back to the control host. Seems like you could just use Jinja2 filters in a playbook to have infinitely-flexible filtering instead of this really limited/obtuse positional filtering baked into the module (bonus, it simplifies the module). Make sense? Or am I still missing something?

fiunchinho commented:
@nitzmahone Most places I've been, launch configurations are named using date times to get unique names and to be able to easily order them. Probably people commeting on this PR follow the same best practice.

I've updated the code to add pagination and I've removed the sort_start and sort_end parameters. I've added a new limit parameter to limit the number of results it returns.

nitzmahone commented on cloud/amazon/ec2_lc_find.py:
```diff
@@ -0,0 +1,226 @@
+    returned: when Launch Configuration was found
+    type: string
+    sample: mykey
+keyname:
```
duplicated docs key, should be security_groups maybe?

nitzmahone commented on cloud/amazon/ec2_lc_find.py:
```diff
@@ -0,0 +1,226 @@
+        region=dict(required=True, aliases=['aws_region', 'ec2_region']),
+        name_regex=dict(required=True),
+        sort_order=dict(required=False, default='ascending', choices=['ascending', 'descending']),
+        limit=dict(required=False),
```
not necessary, but did you know you can avoid the type error checking on this later if you just add type='int' to the argspec dict for this arg?

nitzmahone commented on cloud/amazon/ec2_lc_find.py:
```diff
@@ -0,0 +1,226 @@
+    try:
+        if offset:
+            results = results[:int(offset)]
+    except TypeError:
```
see argspec comment below- you can nuke this check if you just tell us what the type should be

nitzmahone commented:
a couple last comments, and we're good to go

fiunchinho commented:
Good catch! I just pushed again

nitzmahone commented:
Nice- looks like you fixed a couple other doc glitches from your changes I missed in my re-review too... CI is green, merging. Thanks for your patience and quick responses! In the future, feel free to stop by one of our [IRC meetings](https://github.com/ansible/community/blob/master/MEETINGS.md) and hassle us to review your PR if you're frustrated with a lack of community feedback or attention to it.

fiunchinho commented:
I'm really glad it's finally merged. Thank you.

nitzmahone commented:
No problem- it'll ship in Ansible 2.2 September-ish.
